spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -49,16 +55,39 @@ interface ChannelEnd {
   counterpartyChannelIdentifier: Identifier
   connectionHops: [Identifier]
   version: string
+  upgradeSequence: uint64
+  flushStatus: FlushStatus
 }
 ```
 
-The desired property that the channel upgrade protocol MUST NOT modify the underlying clients or channel identifiers, means that only some fields of `ChannelEnd` are upgradable by the upgrade protocol.
+- `state`: The state is specified by the handshake steps of the upgrade protocol and will be mutated in place during the handshake.
+- `upgradeSequence`: The upgrade sequence will be incremented and agreed upon during the upgrade handshake and will be mutated in place.
+
+```typescript
+enum FlushStatus {
+    NOTINFLUSH
+    FLUSHING
+    FLUSHCOMPLETE
+}
+```
+
+FlushStatus will be in `NOTINFLUSH` state when the channel is not in an upgrade handshake. It will be in `FLUSHING` mode when the channel end is flushing in-flight packets. The FlushStatus will change to `FLUSHCOMPLETE` once there are no in-flight packets left and the channelEnd is ready to move to OPEN.
+
+All other parameters will remain the same during the upgrade handshake until the upgrade handshake completes. When the channel is reset to `OPEN` on a successful upgrade handshake, the fields on the channel end will be switched over to the `UpgradeFields` specified in the upgrade.
```

Comment:
Yea wasn't sure exactly how we wanted to handle this. I created 3 states: 

FLUSHING: flushing (during handshake)
FLUSHCOMPLETE: flushing complete (during handshake)
NOTINFLUSH: state when we are not in upgrade handshake

This NOTINFLUSH state is the default enum (0 in the case of golang). I thought it was weird to call it `UNINITIALIZED` since it will be a state we return to after a complete upgrade handshake.

I could also have done 2 states:

`FLUSHED`
`FLUSHING`

Then while packets are processing during normal channel execution, the flush state will be `FLUSHED`; it will only be FLUSHING while we are paused on upgrade handshake to flush packets.

However, I thought this might be confusing so I opted against it.

cc: @colin-axner

Reply:
Open to changing with consensus. cc: @colin-axner

Reply:
I actually like the 3 enums states @AdityaSripal proposed here. To me, it makes sense that `FLUSHCOMPLETE` is a very specific state, it tells the caller that the channel has no inflight packets, while `NOTINFLUSH` or `NOTFLUSHING` just means it isn't blocking on flushing in-flight packets. 

What's the benefit to reducing the number of states?

Reply:
Was discussing with @colin-axner this morning and I had suggested also using a `flushing` boolean value. Either works I guess, I feel like I'd lean slightly towards the boolean.

Reply:
Let's discuss this synchronously

Reply:
ah, the boolean vs enum battle appears once again